### PR TITLE
Sprint3/model development

### PR DESCRIPTION
**Pull request by liav2002**

### CODE REVIEW

File: utils/solver.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+from utils.model_manager import run_models_and_get_best, get_model_instance
+from model.pipeline.pipeline import DataPipeline
+
+
+class Solver:
+    def __init__(self):
+        """
+        Initialize the Solver class with necessary configurations and logger setup.
+        """
+        self.logger = Logger()
```
Comment (DanielGoman) [inline utils/solver.py:16]: So yeah, the logger has to particular reason to be part of this class.

As we discussed, I think it'd be better to just create an instance of the logger in the `utils\logger.py` script, and import this every time.

The idea is just to avoid passing the logger around as an argument to the function, it doesn't have any effect on the logic of the function, so it probably shouldn't be a function argument.


This may not be th ebest approach when dealing with multiple loggers as this may get a bit messy, but for the case of having a single logger for the entire project this should be fine.

An analogy to math woudl be having a functino which has 2 variables, but isn't affected by one of them, something like:

f(x, y) = x^2

No point in having y as a variable as far as I'm concerned 🎋

File: utils/solver.py
Comment (DanielGoman) [inline utils/solver.py:1]: Don't think this file should be in the utils, it literally has the entire logic of the project

File: utils/model_manager.py [hunk 2]
```diff
@@ -3,19 +3,21 @@
...
+        params (dict): Model parameters to initialize the instance.
 
     Returns:
         object: An instance of the specified model class.
     """
     module_name, class_name = model_class_path.rsplit(".", 1)
     module = importlib.import_module(module_name)
-    return getattr(module, class_name)()
+    model_class = getattr(module, class_name)
+    return model_class(**(params or {}))
```
Comment (DanielGoman) [inline utils/model_manager.py:21]: Cool stuff 🐭

File: utils/solver.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        # Define models to evaluate
+        models = {}
+        if self.config["hyperparameter_tuning"]["search_best_params"]:
+            self.logger.log("Performing hyperparameter tuning for all models.")
+            for model_name, model_details in self.config["models"].items():
+                model = get_model_instance(
+                    model_details["class"],
+                    model_details["params"]
+                )
+                if hasattr(model, "search_best_params"):
```
Comment (DanielGoman) [inline utils/solver.py:60]: Is there a reason for this check?

If so, if not having the `search_best_params` method is an error, I think it's proper to log a warning or error

File: utils/solver.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                    self.logger.log(f"Starting hyperparameter tuning for {model_name}.")
+                    best_params = model.search_best_params(
+                        processed_train_df.drop(columns=["user_rating"]),
+                        processed_train_df["user_rating"],
+                        self.config
```
Comment (DanielGoman) [inline utils/solver.py:65]: No need to pass the entire config, you just need the search space of this current model

File: utils/solver.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                for model_name, model_details in self.config["models"].items()
+            }
+
+        self.logger.log("Models initialized.")
+
+        # Run models and find the best model
+        results_df, best_model = run_models_and_get_best(processed_train_df, models, self.config, self.logger)
+        self.logger.log("Model evaluation completed.")
+        self.logger.log("Model Performance Summary:")
+        self.logger.log(results_df.to_string(index=False))
```
Comment (DanielGoman) [inline utils/solver.py:86]: Check the `tabulate` package for pretty printing of a dataframe 🎨

File: utils/solver.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        self.logger.log(f"Best model saved to {self.best_model_path}.")
+
+    def test_best_model(self):
+        """
+        Load the best model, process the test data, and make predictions.
+        """
+        if not os.path.exists(self.best_model_path):
+            self.logger.log(
+                f"Error: Best model file '{self.best_model_path}' not found. Please run 'find_best_model' first.")
+            return
```
Comment (DanielGoman) [inline utils/solver.py:110]: Probably fine to raise an exception here, there remains nothing to do if the best model does not exist.

Correct me if I'm wrong 🦧

File: utils/model_manager.py [hunk 7]
```diff
@@ -3,19 +3,21 @@
 from utils.logger import Logger
 
 
-def get_model_instance(model_class_path: str):
+def get_model_instance(model_class_path: str, params: dict = None):
```
Comment (DanielGoman) [inline utils/model_manager.py:6]: Missing return type

File: model/models/xgboost_model.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
+import optuna
+import pandas as pd
+from typing import Dict
+from xgboost import XGBRegressor
+from sklearn.model_selection import cross_validate
+from model.models.base_model import BaseModel
```
Comment (DanielGoman) [inline model/models/xgboost_model.py:6]: incorrect order of imports

File: model/models/xgboost_model.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+
+class XGBoostModel(BaseModel):
```
Comment (DanielGoman) [inline model/models/xgboost_model.py:9]: So yeah, as we discussed, there's no reason to create a class for each model we run, since the methods they implement (at the moment) are the same for all of them.

This just adds extra work for you, and make the project less generic 👨

File: model/models/xgboost_model.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+        study = optuna.create_study(direction="minimize")
+        study.optimize(objective, n_trials=config["hyperparameter_tuning"]["trials"])
+        self.best_params = study.best_params
+
+        # Update the model with the best parameters
+        self.model.set_params(**self.best_params)
+        return self.best_params
+
+    def get_params(self, deep: bool = True) -> Dict:
```
Comment (DanielGoman) [inline model/models/xgboost_model.py:135]: Is this used?

Is there a case where you don't want it to be a "deep" retrieval of the params?

Comment (DanielGoman) [inline model/models/xgboost_model.py:135]: Generally for getters you could use the `@property` decorator, look into it 👻

File: model/models/xgboost_model.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        subsample_range = search_space["subsample_range"]
+        colsample_bytree_range = search_space["colsample_bytree_range"]
+
+        def objective(trial):
+            # Suggest hyperparameters from the search space
+            max_depth = trial.suggest_int("max_depth", *max_depth_range)
+            learning_rate = trial.suggest_float("learning_rate", *learning_rate_range, log=True)
+            n_estimators = trial.suggest_int("n_estimators", *n_estimators_range)
+            subsample = trial.suggest_float("subsample", *subsample_range)
+            colsample_bytree = trial.suggest_float("colsample_bytree", *colsample_bytree_range)
```
Comment (DanielGoman) [inline model/models/xgboost_model.py:102]: With the proper setup of the config file, you can make a generic function instead of having a separate one for every model

File: model/models/xgboost_model.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+                model,
+                X,
+                y,
+                cv=config["cross_validation"]["cv_folds"],
+                scoring=config["cross_validation"]["scoring"]["rmse"],
+                return_train_score=config["cross_validation"]["return_train_score"]
+            )
+
+            # Return the mean of the test scores (neg RMSE)
+            return -scores["test_score"].mean()
```
Comment (DanielGoman) [inline model/models/xgboost_model.py:125]: What's with the minus? Max of minus the scores is the same as Min over the scores

**Review by DanielGoman — CHANGES_REQUESTED** [review]
friends-joey

File: config/config.yaml
[the remark this answers is not in the document]
Comment (liav2002) [reply config/config.yaml:1]: Got it, next time I'll add a clear and concise description to the PR to ensure the purpose and changes are easy to understand.

[the remark this answers is not in the document]
Comment (liav2002) [reply config/config.yaml:1]: Fixed

[the remark this answers is not in the document]
Comment (liav2002) [reply config/config.yaml:1]: Next time I’ll make sure to follow the proper workflow moving forward to avoid such situations. 👍

File: utils/solver.py (hunk 3 above)
Comment (liav2002) [reply utils/solver.py]: I check this because the `LinearRegression` class don't contains this method.

I didn't implement this for LinearRegression because there are not much interesting hyper parameters for this one, it's a very simple model.

File: model/models/xgboost_model.py (hunk 9 above)
Comment (liav2002) [reply model/models/xgboost_model.py]: I think about that and actually the `evaluate` and `search_params` is a little bit different that's why I prefer to keep these classes.

File: model/models/xgboost_model.py (hunk 10 above)
Comment (liav2002) [reply model/models/xgboost_model.py]: The `get_params` method is part of the scikit-learn estimator interface and is typically required for compatibility with libraries like `GridSearchCV` or `Pipeline`

File: model/models/xgboost_model.py (hunk 11 above)
Comment (liav2002) [reply model/models/xgboost_model.py]: I have implemented a generic ModelHandler class that supports training, evaluation, and hyperparameter tuning for any model, eliminating the need for separate implementations.

File: src/core/solver.py [hunk 13]
```diff
@@ -12,6 +12,7 @@
 
 LOG_DIR = "logs"
 CONFIG_PATH = "config/config.yaml"
+FALSE_ANALYSIS_DIR = "output/false_analysis_df/"
```
Comment (DanielGoman) [inline src/core/solver.py]: consts .-.

File: src/core/solver.py [hunk 14]
```diff
@@ -139,3 +139,47 @@ def test_best_model(self):
...
+
+        # Initialize the pipeline and process the training data
+        logger.log("Initializing data pipeline for false analysis.")
+        pipeline = DataPipeline(self.config["bots_and_scores"], self.turns_df, self.games_df)
+        processed_train_df = pipeline.process_train_data(self.train_df)
+        logger.log("Data pipeline processing for training data completed.")
+
+        # Make predictions on the processed training data
+        logger.log("Making predictions on the training data for false analysis.")
+        predictions = best_model.predict(processed_train_df.drop(columns=["user_rating"]))
```
Comment (DanielGoman) [inline src/core/solver.py:172]: So this is pretty inccorent.

Here you're loading a model which was trained on the entire train set, and evaluating its performance on the train set. Think why that's wrong and what it could lead to.

Also, think what's the proper approach and adjust your False Analysis accordingly 🐱‍👤

File: src/core/solver.py [hunk 15]
```diff
@@ -139,3 +139,47 @@ def test_best_model(self):
...
+        ).abs()
+
+        # Sort by the error column in descending order
+        logger.log("Sorting the dataframe by 'error' in descending order.")
+        sorted_df = processed_train_df.sort_values(by="error", ascending=False)
+
+        # Save the dataframe to the output directory
+        logger.log(f"Saving sorted dataframe to {FALSE_ANALYSIS_DIR}.")
+        os.makedirs(FALSE_ANALYSIS_DIR, exist_ok=True)
+        output_path = os.path.join(FALSE_ANALYSIS_DIR, "false_analysis.csv")
```
Comment (DanielGoman) [inline src/core/solver.py:188]: file name const

File: src/core/solver.py [hunk 16]
```diff
@@ -139,3 +139,47 @@ def test_best_model(self):
...
+        # Make predictions on the processed training data
+        logger.log("Making predictions on the training data for false analysis.")
+        predictions = best_model.predict(processed_train_df.drop(columns=["user_rating"]))
+
+        # Add new columns to the dataframe
+        logger.log("Adding 'predicted_user_rating' and 'error' columns.")
+        processed_train_df["predicted_user_rating"] = predictions
+        processed_train_df["error"] = (
+                processed_train_df["user_rating"] - processed_train_df["predicted_user_rating"]
+        ).abs()
```
Comment (DanielGoman) [inline src/core/solver.py:179]: Whether the model overshot/undershow the prediction has no meaning?

File: src/core/solver.py [hunk 17]
```diff
@@ -57,34 +62,34 @@ def find_best_model(self):
...
-                model_name: get_model_instance(
-                    model_details["class"],
-                    model_details["params"]
+        for model_name, model_details in self.config["models"].items():
+            if model_name == "XGBoost":
+                model_type = XGBRegressor
+            elif model_name == "Linear Regression":
+                model_type = LinearRegression
+            elif model_name == "Neural Network":
+                model_type = MLPRegressor
```
Comment (DanielGoman) [inline src/core/solver.py:73]: Probably should put this as a dict variable in consts

File: src/model/model_handler.py [hunk 18]
```diff
@@ -0,0 +1,138 @@
+import optuna
+import numpy as np
+import pandas as pd
+from typing import Dict, Union
```
Comment (DanielGoman) [inline src/model/model_handler.py:4]: typing should be first, it's builtin, while the rest of the packages are installed

File: src/model/model_handler.py [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+from sklearn.model_selection import cross_validate
+
+
+class ModelHandler:
+    """
+    A generic handler for managing machine learning models, including training, evaluation,
+    and hyperparameter tuning.
+    """
+
+    def __init__(self, model: type, params: Dict[str, Union[int, float, str]]) -> None:
```
Comment (DanielGoman) [inline src/model/model_handler.py:15]: `type` is a function, not an object type, so this annotation is kinda awkward, and now really informative.

This is an interesting point, lets think about it.

You want to annotate that `ModelHandler` accepts a class, but technically you don't know which class.

Surely you have some information about the nature of the classes that can/should be passed to `ModelHandler`.

How do you use this information to better annotate this parameter?

Feel free to hit me up about ideas you have for this 🐱‍👓

**Review by DanielGoman — CHANGES_REQUESTED** [review]
Good work :)
image

File: src/core/solver.py [hunk 20]
```diff
@@ -1,74 +1,83 @@
...
+
+from src.utils.config_loader import load_config
+from src.utils.logger import logger, FileLogger
+from src.utils.data_loader import load_data, load_best_model_path
+from src.model.model_manager import run_models_and_get_best, get_model_instance
+
+from src.model.pipeline.pipeline import DataPipeline
+
+LOG_DIR = "logs"
+CONFIG_PATH = "config/config.yaml"
```
[the remark this answers is not in the document]
Comment (liav2002) [reply src/core/solver.py:19]: Not sure why Not sure why this makes sense.
After all, to read this value I would first have to open the file for reading.
If I already opened it for reading, I probably got the path to it and then the value is meaningless.

File: requirements.txt [hunk 21]
```diff
@@ -6,9 +6,12 @@ xgboost>=1.4.0
...
 PyYAML>=5.4.1
+tabulate>=0.9.0
 
 # Hyperparameter tuning
 optuna>=3.0.0
 
 # Optional dependencies for Neural Network models
 scipy>=1.7.0
+
+DateTime~=5.5
```
[the remark this answers is not in the document]
Comment (liav2002) [reply requirements.txt:17]: I use datetime for the log' files names.

File: src/model/models/neural_network_model.py [hunk 22]
```diff
@@ -1,9 +1,11 @@
 import optuna
 import pandas as pd
 from typing import Dict
```
[the remark this answers is not in the document]
Comment (liav2002) [reply src/model/models/neural_network_model.py]: Why you still see this file?

I remove him, now we have one generic class that works with all the models.

File: config/config.yaml
[the remark this answers is not in the document]
Comment (liav2002) [reply config/config.yaml:1]: Oops..

File: src/core/solver.py (hunk 13 above)
Comment (liav2002) [reply src/core/solver.py]: I move them to seperate file in config/consts.py

File: src/core/solver.py (hunk 14 above)
Comment (liav2002) [reply src/core/solver.py:170]: You're absolutely right, evaluating the model on the training set can lead to overfitting and misleading results. I'll update the False Analysis to use a separate validation set to ensure we're evaluating the model's generalization performance and analyzing real-world errors.

I will do it in separate branch, it's not a part of this one. So, ignore that for now.

File: src/core/solver.py (hunk 15 above)
Comment (liav2002) [reply src/core/solver.py]: Done

File: src/core/solver.py (hunk 16 above)
Comment (liav2002) [reply src/core/solver.py:177]: We will discuss this on separate branch CR.

File: src/core/solver.py (hunk 17 above)
Comment (liav2002) [reply src/core/solver.py]: Done

File: src/model/model_handler.py (hunk 18 above)
Comment (liav2002) [reply src/model/model_handler.py]: Done

File: src/model/model_handler.py [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+        """
+        Initializes the ModelHandler.
+
+        Args:
+            model (type): A machine learning model class (e.g., from scikit-learn, XGBoost, etc.).
+            params (Dict[str, Union[int, float, str]]): Parameters for the model.
```
[the remark this answers is not in the document]
Comment (liav2002) [reply src/model/model_handler.py:21]: I have tool that add it automatically.

File: src/model/model_handler.py (hunk 19 above)
Comment (liav2002) [reply src/model/model_handler.py:15]: You're absolutely right—using type as an annotation isn't very informative. After thinking about it, I realize that the model parameter represents a class that should implement the fit and predict methods (common across scikit-learn, PyTorch, XGBoost, etc.). A better way to annotate this would be to define a Protocol from the typing module that specifies the required methods (fit, predict, and potentially others).

I try to do this.

File: src/model/model_manager.py [hunk 24]
```diff
@@ -34,13 +14,13 @@ def run_models_and_get_best(
 
     Args:
         train_df (pd.DataFrame): Processed training data.
-        models (dict): A dictionary of models to evaluate.
+        models (dict): A dictionary of ModelHandler instances to evaluate.
```
[the remark this answers is not in the document]
Comment (liav2002) [reply src/model/model_manager.py:18]: Done

File: config/consts.py [hunk 25]
```diff
@@ -1,5 +1,11 @@
...
+from sklearn.neural_network import MLPRegressor
+
 LOG_DIR = "logs"
 CONFIG_PATH = "config/config.yaml"
 CONFIG_FILE_PATHS = "config/file_paths.yaml"
 FALSE_ANALYSIS_DIR = "output/false_analysis_df/"
-FALSE_ANALYSIS_FILE = "false_analysis.csv"
\ No newline at end of file
+FALSE_ANALYSIS_FILE = "false_analysis.csv"
+MODEL_2_CLASS_MAP = {"XGBoost": XGBRegressor, "Linear Regression": LinearRegression,
```
Comment (DanielGoman) [inline config/consts.py:10]: Numbers in variables names is kinda sus 🥸

File: src/model/model_handler.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (DanielGoman) [reply src/model/model_handler.py:50]: well alrighty then

File: src/model/model_manager.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (DanielGoman) [reply src/model/model_manager.py:18]:
```suggestion
        models (Dict[str, ModelHandler]): A dictionary of ModelHandler instances to evaluate.
```
Wasn't updated .-.

**Review by DanielGoman — APPROVED** [review]
A couple of small comments, good stuff.

After fixing them feel free to merge :)
